### PR TITLE
feat(vlog): making vlog threshold dynamic 6ce3b7c

### PR DESCRIPTION
**Pull request by aman-bansal**
I ran this for 200M entries. Value threshold is set to 128 initially with max of 2048 and threshold range of 128-1024.
With Dynamic Threshold set to true it takes 52m. Vlog file size were around 4Gi. Threshold sets to around 770 after certain iterations.

With Dynamic Threshold set to false, it takes around 35 min for the write bench to write the data. vlog file size in this case is around 70Gi

Percentage distribution of value sizes were 80% less than 512B, 19% 512B-768B, 0.99% 768-1024B, and 0.01% 1024-2048 B.

This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/badger/1635)

### CODE REVIEW

**Comment by jarifibrahim** [comment]
@aman-bansal can you please resolve the conflicts?

File: structs.go [hunk 1]
```diff
@@ -147,18 +147,34 @@ type Entry struct {
...
+	if e.valThreshold == 0 {
+		e.valThreshold = threshold
 	}
-	return len(e.Key) + 12 + 2 // 12 for ValuePointer, 2 for metas.
+	k := int64(len(e.Key))
+	v := int64(len(e.Value))
+	if v < e.valThreshold {
+		return k + v + 2 // Meta, UserMeta
+	}
+	return k + 12 + 2
```
Comment (NamanJain8) [inline structs.go:167]: Use `vPtrSize` instead of 12. Also, please add back the comment.

File: txn.go [hunk 2]
```diff
@@ -377,8 +377,8 @@ func (txn *Txn) modify(e *Entry) error {
 		return exceedsSize("Key", maxKeySize, e.Key)
 	case int64(len(e.Value)) > txn.db.opt.ValueLogFileSize:
 		return exceedsSize("Value", txn.db.opt.ValueLogFileSize, e.Value)
-	case txn.db.opt.InMemory && len(e.Value) > txn.db.opt.ValueThreshold:
-		return exceedsSize("Value", int64(txn.db.opt.ValueThreshold), e.Value)
+	case txn.db.opt.InMemory && int64(len(e.Value)) > txn.db.valueThreshold():
+		return exceedsSize("Value", txn.db.valueThreshold(), e.Value)
```
Comment (NamanJain8) [inline txn.go:381]: Do we still need this check?

File: value.go [hunk 3]
```diff
@@ -837,6 +842,7 @@ func (vlog *valueLog) write(reqs []*request) error {
 		return nil
 	}
 
+	valueSizes := make(ReqValSize, 0)
```
Comment (NamanJain8) [inline value.go:845]: This needs to be inside the for loop. Also, we can specify the capacity upfront. 
`make(ReqValSiz, 0, len(req.Entries))`

File: value.go [hunk 4]
```diff
@@ -69,6 +69,8 @@ var errDeleteVlogFile = errors.New("Delete vlog file")
 
 type logEntry func(e Entry, vp valuePointer) error
 
+type ReqValSize []int64
```
Comment (NamanJain8) [inline value.go:72]: Do we need a separate type?

File: value.go [hunk 5]
```diff
@@ -1099,3 +1106,74 @@ func (vlog *valueLog) updateDiscardStats(stats map[uint32]int64) {
...
+}
+
+func (v *vlogThreshold) listenForValueThresholdUpdate() {
+	defer v.closer.Done()
+	for {
+		select {
+		case <-v.closer.HasBeenClosed():
+			return
+		case val := <-v.valueCh:
+			if val == nil {
```
Comment (NamanJain8) [inline value.go:1164]: When will this be `nil`?

File: value.go [hunk 6]
```diff
@@ -1099,3 +1106,74 @@ func (vlog *valueLog) updateDiscardStats(stats map[uint32]int64) {
...
+				continue
+			}
+			for _, e := range val {
+				v.vlMetrics.Update(e)
+			}
+			// we are making it to get 99 percentile so that only values
+			// in range of 1 percentile will make it to the value log
```
Comment (NamanJain8) [inline value.go:1171]: Update the comment. We are not making it 99. It depends on `opt.VlogPercentile`.

**Review by NamanJain8 — COMMENTED** [review]
Great @aman-bansal . Got few comments. Also, we are considering the vlog rewrites but not the keys which are deleted while compaction. Should be okay I believe.

File: txn.go (hunk 2 above)
Comment (aman-bansal) [reply txn.go:381]: This check is for inmemory. In this mode, the value threshold will work in a default behaviour. It adds basic validation for that mode i believe.

File: value.go (hunk 4 above)
Comment (aman-bansal) [reply value.go:72]: It was looking bit noisy to add int64 everywhere i.e. channels, req tracker etc. I did this so as to add context to the variable. Any thoughts about this?

File: value.go (hunk 5 above)
Comment (aman-bansal) [reply value.go:1164]: safe check only. No use case per se.

File: value.go (hunk 6 above)
Comment (aman-bansal) [reply value.go]: yes thanks. It was changed recently. comment was outdated thanks :).

**Review by jarifibrahim — COMMENTED** [review]
Got some comments

> Reviewed 1 of 16 files at r4.
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/badger/1635#-:-MS8Hb1e7sWxLAkjPFG6:b-tqxjll)* status: 1 of 18 files reviewed, 12 unresolved discussions (waiting on @aman-bansal, @jarifibrahim, @manishrjain, and @NamanJain8)

___
*[options.go, line 62 at r8](https://reviewable.io/reviews/dgraph-io/badger/1635#-MS8Fw5e9IhCVHoS_RCS:-MS8Fw5e9IhCVHoS_RCT:bd8yvep) ([raw file](https://github.com/dgraph-io/badger/blob/9d1a18d27b94fdb7ab5da9d0cccbadb3d123ce31/options.go#L62)):*
> ```Go
> 
> 	VLogPercentile    float64
> 	MinValueThreshold int64
> ```

This is a breaking change. This PR doesn't need to be a breaking change.
___
*[options.go, line 348 at r8](https://reviewable.io/reviews/dgraph-io/badger/1635#-MS8Eyzl2fxj5_4SUXKE:-MS8Eyzl2fxj5_4SUXKF:b78bmdc) ([raw file](https://github.com/dgraph-io/badger/blob/9d1a18d27b94fdb7ab5da9d0cccbadb3d123ce31/options.go#L348)):*
> ```Go
> // [MinValueThreshold, Options.maxValueThreshold]
> //
> // Say VLogPercentile with 1.0 means threshold will eventually sets too Options.maxValueThreshold
> ```

will be eventually set to ...
___
*[structs.go, line 172 at r8](https://reviewable.io/reviews/dgraph-io/badger/1635#-MS8GSHB8_bg8wJ0nkMd:-MS8GSHB8_bg8wJ0nkMe:b-xitzfx) ([raw file](https://github.com/dgraph-io/badger/blob/9d1a18d27b94fdb7ab5da9d0cccbadb3d123ce31/structs.go#L172)):*
> ```Go
> func (e *Entry) skipVlog(threshold int64) bool {
> 	if e.valThreshold == 0 {
> 		e.valThreshold = threshold
> ```

When would valThreshold be zero? This function doesn't need to change the value of the `valThreshold`.
___
*[structs.go, line 174 at r8](https://reviewable.io/reviews/dgraph-io/badger/1635#-MS8HN0b2LCNDyf7BG2i:-MS8HN0b2LCNDyf7BG2j:b5fmgdu) ([raw file](https://github.com/dgraph-io/badger/blob/9d1a18d27b94fdb7ab5da9d0cccbadb3d123ce31/structs.go#L174)):*
> ```Go
> 		e.valThreshold = threshold
> 	}
> 	if int64(len(e.Value)) < e.valThreshold {
> ```

`return int64(len(e.Value)) < e.valThreshold `
___
*[value.go, line 1164 at r7](https://reviewable.io/reviews/dgraph-io/badger/1635#-MQI5EP-it6wpth42BlK-r7-1164:-MS8FVrz577ifR192IGL:b-v8u2np) ([raw file](https://github.com/dgraph-io/badger/blob/b2a80eed88d51f9b468d3685499398d86475968c/value.go#L1164)):*
<details><summary><i>Previously, aman-bansal (aman bansal) wrote…</i></summary><blockquote>

safe check only. No use case per se.
</blockquote></details>

If we don't expect this to be nil, I'd at least log it.
___
*[value.go, line 892 at r8](https://reviewable.io/reviews/dgraph-io/badger/1635#-MS8GtpD6Zf5gwrVzsoH:-MS8GtpD6Zf5gwrVzsoI:bh3tkuz) ([raw file](https://github.com/dgraph-io/badger/blob/9d1a18d27b94fdb7ab5da9d0cccbadb3d123ce31/value.go#L892)):*
> ```Go
> 
> 		vlog.numEntriesWritten += uint32(written)
> 		vlog.db.threshold.update(valueSizes)
> ```

@aman-bansal do you need to update the sizes one txn at a time? Can't we do this outside both the for loops? One single call to `threshold.update(...)`
___
*[value.go, line 1175 at r8](https://reviewable.io/reviews/dgraph-io/badger/1635#-MS8FdwsE7qGp1z9AWkU:-MS8FdwsE7qGp1z9AWkV:b-el7ntx) ([raw file](https://github.com/dgraph-io/badger/blob/9d1a18d27b94fdb7ab5da9d0cccbadb3d123ce31/value.go#L1175)):*
> ```Go
> 			p := int64(v.vlMetrics.Percentile(v.percentile))
> 			if atomic.LoadInt64(&v.valueThreshold) != p {
> 				// v.opt.Infof("updating value threshold from: %d to: %d", vt, p)
> ```

stale comment.

**Review by jarifibrahim — APPROVED** [review]
Reset the histogram on drop all otherwise LGTM.

**Review by manishrjain — COMMENTED** [review]
Hey @aman-bansal , what's the status here?

> *[Reviewable](https://reviewable.io/reviews/dgraph-io/badger/1635#-:-MSdKidIBuObj4pYX0Vd:bz4llug)* status: 1 of 18 files reviewed, 13 unresolved discussions (waiting on @aman-bansal, @jarifibrahim, and @NamanJain8)

___
*[db.go, line 119 at r4](https://reviewable.io/reviews/dgraph-io/badger/1635#-MR_r5UfCrzXw1ZJEUMf:-MR_r5UfCrzXw1ZJEUMg:bdzwr8r) ([raw file](https://github.com/dgraph-io/badger/blob/2230255f90434d0bda0a1ec50fc3302d5cd1a027/db.go#L119)):*
> ```Go
> 	opt.maxBatchSize = (15 * opt.MemTableSize) / 100
> 	opt.maxBatchCount = opt.maxBatchSize / int64(skl.MaxNodeSize)
> 	//opt.maxValueThreshold = 1024.0
> ```

Rename opt.ValueThreshold to opt.MinValueThreshold. And then this one is max. And do add a comment.

**Review by aman-bansal — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/badger/1635)* status: 1 of 15 files reviewed, 13 unresolved discussions (waiting on @jarifibrahim, @manishrjain, and @NamanJain8)

___
*[db.go, line 119 at r4](https://reviewable.io/reviews/dgraph-io/badger/1635#-MR_r5UfCrzXw1ZJEUMf:-MSgsZ9c1aboj4vNaFGu:bfo0351) ([raw file](https://github.com/dgraph-io/badger/blob/2230255f90434d0bda0a1ec50fc3302d5cd1a027/db.go#L119)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Rename opt.ValueThreshold to opt.MinValueThreshold. And then this one is max. And do add a comment.
</blockquote></details>

Done. We have kept the name same because otherwise it would become breaking change. we believe this change doesnt need to be breaking change.
___
*[options.go, line 62 at r8](https://reviewable.io/reviews/dgraph-io/badger/1635#-MS8Fw5e9IhCVHoS_RCS:-MSgsmW65GOuKa7_kYaC:b-896fix) ([raw file](https://github.com/dgraph-io/badger/blob/9d1a18d27b94fdb7ab5da9d0cccbadb3d123ce31/options.go#L62)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

This is a breaking change. This PR doesn't need to be a breaking change.
</blockquote></details>

Done.
___
*[options.go, line 348 at r8](https://reviewable.io/reviews/dgraph-io/badger/1635#-MS8Eyzl2fxj5_4SUXKE:-MSgsoTX5UFPQHplIL-g:b-896fix) ([raw file](https://github.com/dgraph-io/badger/blob/9d1a18d27b94fdb7ab5da9d0cccbadb3d123ce31/options.go#L348)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

will be eventually set to ...
</blockquote></details>

Done.
___
*[structs.go, line 167 at r7](https://reviewable.io/reviews/dgraph-io/badger/1635#-MQI5EP-it6wpth42BlI-r7-167:-MSgsp_92mV9tixMNlPE:b-896fix) ([raw file](https://github.com/dgraph-io/badger/blob/b2a80eed88d51f9b468d3685499398d86475968c/structs.go#L167)):*
<details><summary><i>Previously, NamanJain8 (Naman Jain) wrote…</i></summary><blockquote>

Use `vPtrSize` instead of 12. Also, please add back the comment.
</blockquote></details>

Done.
___
*[structs.go, line 172 at r8](https://reviewable.io/reviews/dgraph-io/badger/1635#-MS8GSHB8_bg8wJ0nkMd:-MSgsrfkA6Cz94VItjx8:b-896fix) ([raw file](https://github.com/dgraph-io/badger/blob/9d1a18d27b94fdb7ab5da9d0cccbadb3d123ce31/structs.go#L172)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

When would valThreshold be zero? This function doesn't need to change the value of the `valThreshold`.
</blockquote></details>

Done.
___
*[structs.go, line 174 at r8](https://reviewable.io/reviews/dgraph-io/badger/1635#-MS8HN0b2LCNDyf7BG2i:-MSgssxB4XAOZcGGo_uF:b-896fix) ([raw file](https://github.com/dgraph-io/badger/blob/9d1a18d27b94fdb7ab5da9d0cccbadb3d123ce31/structs.go#L174)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

`return int64(len(e.Value)) < e.valThreshold `
</blockquote></details>

Done.
___
*[txn.go, line 381 at r7](https://reviewable.io/reviews/dgraph-io/badger/1635#-MQI5EP-it6wpth42BlJ-r7-381:-MSgstwRAQm321RG0-iJ:b-896fix) ([raw file](https://github.com/dgraph-io/badger/blob/b2a80eed88d51f9b468d3685499398d86475968c/txn.go#L381)):*
<details><summary><i>Previously, aman-bansal (aman bansal) wrote…</i></summary><blockquote>

This check is for inmemory. In this mode, the value threshold will work in a default behaviour. It adds basic validation for that mode i believe.
</blockquote></details>

Done.
___
*[value.go, line 72 at r7](https://reviewable.io/reviews/dgraph-io/badger/1635#-MQI5EP-it6wpth42BlK-r7-72:-MSgsuQ44cc_1DM8qpGy:b-896fix) ([raw file](https://github.com/dgraph-io/badger/blob/b2a80eed88d51f9b468d3685499398d86475968c/value.go#L72)):*
<details><summary><i>Previously, aman-bansal (aman bansal) wrote…</i></summary><blockquote>

It was looking bit noisy to add int64 everywhere i.e. channels, req tracker etc. I did this so as to add context to the variable. Any thoughts about this? 
</blockquote></details>

Done.
___
*[value.go, line 845 at r7](https://reviewable.io/reviews/dgraph-io/badger/1635#-MQI5EP-it6wpth42BlK-r7-845:-MS8JU8b4Ni0vaoCbdyz:b-896fix) ([raw file](https://github.com/dgraph-io/badger/blob/b2a80eed88d51f9b468d3685499398d86475968c/value.go#L845)):*
<details><summary><i>Previously, NamanJain8 (Naman Jain) wrote…</i></summary><blockquote>

This needs to be inside the for loop. Also, we can specify the capacity upfront. 
`make(ReqValSiz, 0, len(req.Entries))`
</blockquote></details>

Done.
___
*[value.go, line 1164 at r7](https://reviewable.io/reviews/dgraph-io/badger/1635#-MQI5EP-it6wpth42BlK-r7-1164:-MSgt--680bvfvjmNB5E:b-896fix) ([raw file](https://github.com/dgraph-io/badger/blob/b2a80eed88d51f9b468d3685499398d86475968c/value.go#L1164)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

If we don't expect this to be nil, I'd at least log it.
</blockquote></details>

Done.
___
*[value.go, line 1171 at r7](https://reviewable.io/reviews/dgraph-io/badger/1635#-MQI5EP-it6wpth42BlK-r7-1171:-MS8JIoK2hTe9YJlpDXf:b-896fix) ([raw file](https://github.com/dgraph-io/badger/blob/b2a80eed88d51f9b468d3685499398d86475968c/value.go#L1171)):*
<details><summary><i>Previously, aman-bansal (aman bansal) wrote…</i></summary><blockquote>

yes thanks. It was changed recently. comment was outdated thanks :).
</blockquote></details>

Done.
___
*[value.go, line 892 at r8](https://reviewable.io/reviews/dgraph-io/badger/1635#-MS8GtpD6Zf5gwrVzsoH:-MSgsyZWBXmiaxAZK6q_:b-896fix) ([raw file](https://github.com/dgraph-io/badger/blob/9d1a18d27b94fdb7ab5da9d0cccbadb3d123ce31/value.go#L892)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

@aman-bansal do you need to update the sizes one txn at a time? Can't we do this outside both the for loops? One single call to `threshold.update(...)`
</blockquote></details>

Done.
___
*[value.go, line 1175 at r8](https://reviewable.io/reviews/dgraph-io/badger/1635#-MS8FdwsE7qGp1z9AWkU:-MS8JHEKEFC0WkHpWw9d:b-896fix) ([raw file](https://github.com/dgraph-io/badger/blob/9d1a18d27b94fdb7ab5da9d0cccbadb3d123ce31/value.go#L1175)):*
<details><summary><i>Previously, jarifibrahim (Ibrahim Jarif) wrote…</i></summary><blockquote>

stale comment.
</blockquote></details>

Done.

**Pull request merged**